### ISSUE TITLE
Configurable tab-size for project/repo UI

### ISSUE THREAD

**Issue by maximal** [isaacs/github#1490]
Depending on complexity and/or nesting level various projects need to have different tab sizes on file and diff pages.

I think it should not be constant (as described in #170) and repository level configuration would be just fine.
And it is better to have it not 8-space size (standard value in most browsers) by default.

**Comment by maximal** [comment]
@ioquatix, of course, it would be much better to have full `.editorconfig` support, but I think it is harder to implement.

**Comment by Alhadis** [comment]
@ioquatix GitHub currently respects the `indent_size` rule declared in a repository's `.editorconfig` file, but this is a kludgy and inadequate way of addressing the issue.

For example, authors who use use tabs considerably and semantically would prefer users to read their source code at *their* preferred tabstop width (the whole argument for using tabs in the first place...). This is impossible because GitHub are forcing users to set an `indent_size` to improve the appearance of rendered source code, at the expense of shoving a wanton preference onto contributors working with their code.

**Comment by ioquatix** [comment]
Totally agree. I've argued this point before but seems like very little progress has been made.

**Comment by Alhadis** [comment]
The default value should be left at 8. Many codebases — particularly older or historical ones — have codebases which assume `\t` will always be a multiple of 8.

But we need a repository-level setting. The fact GitHub are ignoring such an obvious (and *trivial*) feature suggests they're actively hurting the tabs-vs-spaces debate by turning a blind eye. Since they're all Ruby programmers, it wouldn't surprise me.

I've requested this many, many, *many* times from site support, and over the course of many years. Nothing. Nothing, nothing, nothing. It's infuriating.

**Comment by ioquatix** [comment]
Here is the email I sent them:

> I believe that you should add some new preferences:
> 
> 1/ A per-user/organisation tab-width preference which applies to all code within that user/organisation.
> 
> 2/ A per-project tab-width preference, which overrides the previously mentioned one.
> 
> 3/ And finally, a per-user tab-width preference which overrides 1/ above but not 2/, which applies to code the user is viewing.
> 
> 4/ If an .editorconfig file exists, it overrides everything for that project.
> 
> The current state w.r.t. tabs in source code on GitHub is pretty bad, tabs with width = 8 is not ideal.
> 
> Additionally, 3/ above could have a min/max for mobile/desktop viewing, respectively - this is very nice because tab width can be 2 on a mobile browser and 4 on a desktop.

**Comment by maximal** [comment]
If you feel it like that, please consider upvoting similar GitLab issue:

https://gitlab.com/gitlab-org/gitlab-ce/issues/57758

**Comment by Alhadis** [comment]
@maximal I feel strongly enough about this to register an account on GitLab just to [voice support](https://gitlab.com/gitlab-org/gitlab-ce/issues/57758#note_141666553) for this.

**Comment by maximal** [comment]
@Alhadis, you’re awesome!

**Comment by Alhadis** [comment]
I'll just leave this here...

https://github.com/Alhadis/Unfuck-Tabstops

Just in case any site staff are convinced trying to push spaces onto staunch tab-users is a good idea.

(it currently needs fixing, will get to that now that I'm reminded... 😀)

**Comment by ioquatix** [comment]
There are some corresponding issues on GitLab: https://gitlab.com/gitlab-org/gitlab-ce/issues/57758
